### PR TITLE
AP_BattMonitor: Allow SMBus batteries to not have cell voltages

### PR DESCRIPTION
**Pull request by WickedShell**
Despite @rmackay9's [comments](https://github.com/ArduPilot/ardupilot/pull/6351#issuecomment-306383952) about not really caring about checking for cell voltages actually being present in the Maxell driver I think this definitely matters, especially for some planned use cases for this helper functions. (On my near term to implement list is failsafes based on cell voltage spread, which I would be using this helper to determine if the failsafe is even possible to check for).

Anyways the reason this check is needed is there is no guarantee that a SMBus battery will actually respond to the requested cell voltages as they are fetched from registers that aren't a part of the formal SMBus spec (although it does appear to be part of the defacto one)

Also took the moment to tag myself as the SMBus battery maintainer, if that's alright with others .

### CODE REVIEW

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h [hunk 1]
```diff
@@ -23,8 +23,7 @@ class AP_BattMonitor_SMBus : public AP_BattMonitor_Backend
     // virtual destructor to reduce compiler warnings
     virtual ~AP_BattMonitor_SMBus() {}
 
-    // all smart batteries provide current and individual cell voltages
-    bool has_current() const override { return true; }
+    // all smart batteries are expected to provide current
     bool has_cell_voltages() const override { return true; }
```
Comment (OXINARF) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h]: You removed the wrong one. Here is what you need: :sleeping_bed: :sleeping:
 :smile:

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Maxell.h [hunk 2]
```diff
@@ -15,6 +15,10 @@ class AP_BattMonitor_SMBus_Maxell : public AP_BattMonitor_SMBus
                              AP_BattMonitor::BattMonitor_State &mon_state,
                              AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev);
 
+    // generic SMBus batteries may not always report the cell voltages, as we
+    // the data is being read from manufacturer specific fields
```
Comment (OXINARF) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Maxell.h]: as we the data -> as the data

I also disagree with what is written here. This isn't a generic driver, it is the Maxell driver, so we know it reports the cell voltages.

**Comment by OXINARF** [comment]
By the way, I'm more and more convinced that doing +1 on the log is wrong. If it goes to UINT16_MAX then it much more clearly shows a reading failure - besides the Solo driver not ever putting UINT16_MAX in the cell voltages.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Maxell.h (hunk 2 above)
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Maxell.h]: I don't think that's true (I'm happy to rename the drivers to be more clear).

The "Maxell" driver only implements SMBus Battery Data Specifications with the one add on of battery cell voltages being read from a combination of OptionalMfGFunctions and reserved fields. However this one add on is also commonly used by other manufacturers (Tattu among others), but for other people like TI chips it has a very different meaning. Now there is an argument we should have a parameter that lets users not query this, but this is the only SMBus driver we have as an option for people if they aren't using a solo.

The solo does current and cell voltages in it's own non standard way which makes it useless for any other smart battery.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h (hunk 1 above)
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus.h]: Oh jeez, you saw nothing....

**Comment by WickedShell** [comment]
@OXINARF For the record I also think the +1 is nasty, it was needed because it was making it really hard for most people to graph the data (IE when @Pedals2Paddles was trying to graph cell voltages it would go to the maximum value for a couple of log messages before we were pulling data from the battery). The reason that logging 0 as a non present cell doesn't work for me, is depending on how your battery is setup it is technically possible that a cell in the middle of the pack could go to 0 but the device is still functioning. There is an argument that can be made that states this last concern is pointless because whoever is reviewing the data will know that if they had a battery with 5 cells and the 3rd one is reporting 0 or max value it's dead...

If we make solo not output the cell voltage logs until we've actually gotten an anwser then I'm happy to ditch the +1 at that point. Regarding solo, we know that the connected battery spec will always implement the cell voltages query which is why I was fine with just always returning true from it.

**Comment by WickedShell** [comment]
Actually, how about this, you don't like the peeking into the first cell voltage only, what if I just add a _has_recieved_cell_voltages bool to AP_BattMonitor_SMBus, then in both the maxell and solo drivers whenever any cell voltage query succeeds I mark that as true, and we always just return the bool inside the AP_BattMonitor_SMBus header? Might be a bit cleaner...

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Maxell.h (hunk 2 above)
Comment (OXINARF) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Maxell.h]: I was the first person saying that I desired a common SMBus driver and I still think that can be achieved by determining in run-time what battery it is.
But I don't understand the difference you make between the Solo and Maxell drivers. You say that Maxell uses one method that isn't in the standard spec and Solo uses two, how does that make them different? One method is the difference between being a generic driver and being useless to any other chip maker? The only thing strange in Solo spec is the current register not being the standard, but other than that it is as generic as Maxell (that, in my opinion, isn't).

I think where we disagree is that you are seeing the Solo battery as the only battery that will ever use those two methods, while the Maxell one will be used in more. But I don't see it that way, you might have not found the same chip that is in Solo batteries in other ones, but it doesn't mean it doesn't or won't exist.
Also, as this market starts to expand, we will probably see more chips and other non-standard methods will appear.

**Comment by OXINARF** [comment]
> Actually, how about this, you don't like the peeking into the first cell voltage only, what if I just add a _has_recieved_cell_voltages bool to AP_BattMonitor_SMBus, then in both the maxell and solo drivers whenever any cell voltage query succeeds I mark that as true, and we always just return the bool inside the AP_BattMonitor_SMBus header? Might be a bit cleaner...

That looks acceptable to me.

**Comment by rmackay9** [comment]
I like that last idea of using the boolean if any cell voltage query succeeds.  In case my opinion is holding things up (which I don't think it is) I'm happy with this going in whenever you two are happy.

**Comment by WickedShell** [comment]
I need to get the time and do the conversion to the bool approach, I'm also of the opinion it's better, I just didn't have the idea originally :)

**Comment by WickedShell** [comment]
The changes were all done over a month ago, is this waiting on anything or has it just slipped off the radar? (I'd forgotten about it actually which is why I hadn't nudged anyone yet).

**Comment by OXINARF** [comment]
Simply because you didn't comment, so I never knew it was updated 🙂 Merged, thanks!

**Pull request merged**